src/webdav/NullResource.py

```diff
@@ -111,6 +112,15 @@ def _default_PUT_factory(self, name, typ, body):
             ob = ZopePageTemplate(name, body, content_type=typ)
         elif typ[:6] == 'image/':
             ob = Image(name, '', body, content_type=typ)
+        elif typ == 'application/octet-stream':
+            if ext in ('.jpg', '.jpeg', '.gif', '.png', '.tiff'):
+                ob = Image(name, '', body)
+            elif ext in ('.xml',):
+                ob = ZopePageTemplate(name, body, content_type='text/xml')
+            elif ext in ('.pt', '.zpt', '.html', '.htm'):
+                ob = ZopePageTemplate(name, body, content_type='text/html')
+            else:
+                ob = File(name, '', body, content_type=typ)
```

Comment:
I do not like this part:
* if the client does not have content type related evidence it should not send content type information; in this case we would guess the content type (and, of course, may be wrong); if the client sends content type information, we should use it and do not guess
* I would not special case some image suffixes but instead delegate to `guess_content_type`; if this guesses some `image` type, I would create an `Image`; special casing some pure Zope extensions is okay

Reply:
I have changed the code to just call `guess_content_type` if no content type information exists or if it is the generic `application/octet-stream` to cover Georg's broken client